dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readDelim.kt

```diff
@@ -0,0 +1,343 @@
+ * @include [QUOTE]
+ * @include [IGNORE_SURROUNDING_SPACES]
+ * @include [TRIM_INSIDE_QUOTED]
+ * @include [PARSE_PARALLEL]
+ * @include [ADJUST_CSV_SPECS]
+ */
+internal fun readDelimImpl(
+    inputStream: InputStream,
+    delimiter: Char,
+    header: List<String>,
+    hasFixedWidthColumns: Boolean,
+    fixedColumnWidths: List<Int>,
+    compression: Compression<*>,
+    colTypes: Map<String, ColType>,
+    skipLines: Long,
+    readLines: Long?,
+    parserOptions: ParserOptions,
+    ignoreEmptyLines: Boolean,
+    allowMissingColumns: Boolean,
+    ignoreExcessColumns: Boolean,
+    quote: Char,
+    ignoreSurroundingSpaces: Boolean,
+    trimInsideQuoted: Boolean,
+    parseParallel: Boolean,
+    adjustCsvSpecs: AdjustCsvSpecs,
+): DataFrame<*> {
+    // set up the csv specs
+    val csvSpecs = with(CsvSpecs.builder()) {
+        customDoubleParser(DataFrameCustomDoubleParser(parserOptions))
+        nullValueLiterals(parserOptions.nullStrings ?: DEFAULT_NULL_STRINGS)
+        headerLegalizer(::legalizeHeader)
+        numRows(readLines ?: Long.MAX_VALUE)
+        ignoreEmptyLines(ignoreEmptyLines)
+        allowMissingColumns(allowMissingColumns)
+        ignoreExcessColumns(ignoreExcessColumns)
+        if (!hasFixedWidthColumns) delimiter(delimiter)
+        quote(quote)
+        ignoreSurroundingSpaces(ignoreSurroundingSpaces)
+        trim(trimInsideQuoted)
+        concurrent(parseParallel)
+        header(header)
+        hasFixedWidthColumns(hasFixedWidthColumns)
+        if (hasFixedWidthColumns && fixedColumnWidths.isNotEmpty()) fixedColumnWidths(fixedColumnWidths)
+        skipLines(takeHeaderFromCsv = header.isEmpty(), skipLines = skipLines)
+
+        // Deephaven's LocalDateTime parser is unconfigurable, so if the user provides a locale, pattern, or formatter,
+        // we must use our own parser for LocalDateTime and let Deephaven read them as Strings.
+        val useDeepHavenLocalDateTime = with(parserOptions) {
+            locale == null && dateTimePattern == null && dateTimeFormatter == null
+        }
+        parsers(parserOptions, colTypes, useDeepHavenLocalDateTime)
+
+        adjustCsvSpecs(this, this)
+    }.build()
+
+    val csvReaderResult = inputStream.useDecompressed(compression) { decompressedInputStream ->
+        // read the csv
+        try {
+            @Suppress("ktlint:standard:comment-wrapping")
+            CsvReader.read(
+                /* specs = */ csvSpecs,
+                /* stream = */ decompressedInputStream.skippingBomCharacters(),
+                /* sinkFactory = */ ListSink.SINK_FACTORY,
+            )
+        } catch (e: CsvReaderException) {
+            // catch case when the file is empty and header needs to be inferred from it.
+            if (e.message ==
+                "Can't proceed because hasHeaderRow is set but input file is empty or shorter than skipHeaderRows"
+            ) {
+                return@readDelimImpl DataFrame.empty()
```

Comment:
Are you sure about it? I incline to think that this is an error that needs to be fixed, having and empty dataframe instead of the exception will be not-so-clear

Reply:
It was to align with the old behavior: feeding an empty file/string should result in an empty dataframe (we could change this of course).

Deephaven can do this successfully if you supply a header yourself (it simply returns empty columns), however it throws an exception when it needs to guess a header from an empty file, like here. But if the file is empty anyway, it makes no sense to require the user to supply a header to avoid this exception, so that's why I caught it like this.

I also tried detecting whether the inputstream is empty before passing it to CsvReader, but I didn't find a reliable way (`.available()` can return `0`, even though the stream still has data).